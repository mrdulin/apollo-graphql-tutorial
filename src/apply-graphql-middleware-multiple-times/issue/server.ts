import { ApolloServer } from 'apollo-server';
import { schema } from './schema';

const server = new ApolloServer({ schema });
const port = 3000;
server.listen(port).then(({ url }) => console.log(`Server is ready at ${url}`));
