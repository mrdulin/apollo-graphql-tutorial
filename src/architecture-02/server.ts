import { ApolloServer } from 'apollo-server-express';
import { DataSources } from 'apollo-server-core/dist/graphqlOptions';
import express from 'express';
import http from 'http';

import {
  schemaWithMiddleware,
  AddressDataSourceImpl,
  UserDataSourceImpl,
  IAddressDataSource,
  IUserDataSource,
} from './schema';
import { ContextFunction } from 'apollo-server-core';
import { credentials } from './credentials';
import { IPostDataSource, PostDataSourceImpl } from './modules/post';
import { logger } from '../util';

interface IAppContext {
  dataSources: {
    address: IAddressDataSource;
    user: IUserDataSource;
    post: IPostDataSource;
  };
}

interface IServerOptions {
  PORT: string | number;
  GRAPHQL_ENDPOINT: string;
}

async function createServer(options: IServerOptions): Promise<http.Server> {
  const { PORT, GRAPHQL_ENDPOINT } = options;
  const app = express();
  const httpServer = http.createServer(app);

  const contextFunction: ContextFunction = ({ req }) => {
    return {
      request: req,
    };
  };

  const apolloServer = new ApolloServer({
    schema: schemaWithMiddleware,
    introspection: true,
    dataSources: (): DataSources<IAppContext> => ({
      address: new AddressDataSourceImpl(),
      user: new UserDataSourceImpl(),
      post: new PostDataSourceImpl(),
    }),
    context: contextFunction,
    engine: {
      apiKey: credentials.APOLLO_ENGINE_API_KEY,
    },
    tracing: true,
  });

  apolloServer.applyMiddleware({ app, path: GRAPHQL_ENDPOINT });

  return new Promise((resolve, reject) => {
    httpServer
      .listen(PORT)
      .on('listening', () => {
        logger.info(`ðŸš€ Http Server ready at http://localhost:${PORT}`);
        logger.info(`ðŸš€ GraphQL Server ready at http://localhost:${PORT}${GRAPHQL_ENDPOINT}`);
        resolve(httpServer);
      })
      .on('error', (error) => {
        reject(error);
      });
  });
}

export { IAppContext, createServer };
