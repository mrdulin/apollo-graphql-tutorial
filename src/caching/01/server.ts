import express from 'express';
import { ApolloServer, gql } from 'apollo-server-express';
import responseCachePlugin from 'apollo-server-plugin-response-cache';
import faker from 'faker';

const typeDefs = gql`
  type User {
    name: String
    email: String
  }
  type Query {
    user: User
  }
`;

const resolvers = {
  Query: {
    user: async (_, __, context, { cacheControl }) => {
      console.count('resolve user');
      cacheControl.setCacheHint({ maxAge: 40 });
      const user = { name: faker.name.findName(), email: faker.internet.exampleEmail() };
      return user;
    },
  },
};

const app = express();
const port = 3001;
const graphqlPath = '/graphql';
const server = new ApolloServer({
  typeDefs,
  resolvers,
  tracing: true,
  // use LRU-memory cache by default
  plugins: [responseCachePlugin()],
  // set cache-control HTTP response header as expected
  // cacheControl: {
  //   defaultMaxAge: 5,
  // },
  // bug? DOES NOT set cache-control HTTP response header
  cacheControl: true,
});

server.applyMiddleware({ app, path: graphqlPath });

app.get('/', (req, res) => {
  res.sendFile('index.html', { root: __dirname });
});

if (require.main === module) {
  app.listen(port, () => {
    console.log(`Apollo server is listening on http://localhost:${port}${graphqlPath}`);
  });
}
