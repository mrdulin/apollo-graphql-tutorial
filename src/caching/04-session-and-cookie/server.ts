import express from 'express';
import { ApolloServer, gql, AuthenticationError } from 'apollo-server-express';
import { RedisCache } from 'apollo-server-cache-redis';
import faker from 'faker';
import responseCachePlugin from 'apollo-server-plugin-response-cache';
import jwt from 'jsonwebtoken';
import cookieParser from 'cookie-parser';
import { GraphQLError } from 'graphql';

interface ISession {
  id: string | number;
  userId: string | number;
  ipAddress: string;
  userAgent: string;
}

interface IUser {
  id: string | number;
  email: string;
  password: string;
}

interface IDB {
  sessions: ISession[];
  users: IUser[];
}

const memoryDb: IDB = {
  sessions: [],
  users: [
    { id: 1, email: 'lin@gmail.com', password: '123' },
    { id: 2, email: 'du@gmail.com', password: 'abc' },
  ],
};
const jwtSecret = 'test secret';

const typeDefs = gql`
  type User @cacheControl(maxAge: 30, scope: PRIVATE) {
    id: ID
    email: String
  }
  type Query {
    user: User
  }
  type Response {
    success: Boolean
  }
  type Mutation {
    login(email: String!, password: String!): Response
  }
`;

const resolvers = {
  Query: {
    user: async (_, __, { db, req }) => {
      console.count('resolve user');
      const sessionToken = req.cookies.sId;
      if (!sessionToken) {
        throw new AuthenticationError('Unauthenticated');
      }
      let tokenDecrypted;
      try {
        tokenDecrypted = jwt.verify(sessionToken, jwtSecret);
      } catch (error) {
        console.log(error);
        throw new GraphQLError('Internal server error');
      }
      const { sId } = tokenDecrypted;
      if (!sId) {
        throw new AuthenticationError('Unauthenticated');
      }
      const session = db.sessions.find((s) => s.id === sId);
      const user = db.users.find((u) => u.id === session.userId);
      if (!user) {
        throw new AuthenticationError(`user not found`);
      }
      return user;
    },
  },
  Mutation: {
    login: async (_, { email, password }, { req, res, db }) => {
      const userAgent = req.get('user-agent');
      const ipAddress = req.get('x-forwarded-for') || req.connection.remoteAddress || req.socket.remoteAddress;
      console.log(userAgent, ipAddress);
      // query user
      const user = db.users.find((u) => u.email === email);
      if (!user) {
        throw new AuthenticationError(`user not found for email: ${email}`);
      }
      if (user.password !== password) {
        throw new AuthenticationError('password wrong');
      }
      // create session
      const session = { id: faker.random.uuid(), userId: user.id, ipAddress, userAgent };
      db.sessions.push(session);
      // create session id
      const sId = jwt.sign({ sId: session.id }, jwtSecret);
      // set cookie
      res.cookie('sId', sId, { maxAge: 60 * 1000 });
      return { success: true };
    },
  },
};

const app = express();
const port = 3001;
const graphqlPath = '/graphql';

function context({ req, res }) {
  return { req, res, db: memoryDb };
}

const server = new ApolloServer({
  typeDefs,
  resolvers,
  plugins: [
    responseCachePlugin({
      sessionId: (requestContext) => {
        return requestContext.context.req.cookies.sId;
      },
    }),
  ],
  context,
  cache: new RedisCache({
    port: 6379,
    host: '127.0.0.1',
    family: 4,
    db: 0,
  }),
});

app.use(cookieParser());
server.applyMiddleware({ app, path: graphqlPath });

app.get('/', (req, res) => {
  res.sendFile('index.html', { root: __dirname });
});

if (require.main === module) {
  app.listen(port, () => {
    console.log(`Apollo server is listening on http://localhost:${port}${graphqlPath}`);
  });
}
