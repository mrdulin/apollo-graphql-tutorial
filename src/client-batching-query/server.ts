import { ApolloServer, ServerInfo } from 'apollo-server';
import { typeDefs } from './typeDefs';
import { resolvers } from './resolver';
import { memoryDB } from './db/memoryDB';
import { ApolloServerBase } from 'apollo-server-core';

async function createApolloServer(): Promise<ApolloServerBase> {
  const PORT = process.env.PORT || 3000;
  const server = new ApolloServer({ typeDefs, resolvers, context: { db: memoryDB } });

  await server
    .listen(PORT)
    .then(({ url }: ServerInfo) => {
      console.log(`ðŸš€ Server ready at ${url}`);
    })
    .catch((error) => {
      console.error('Create server failed.');
      console.error(error);
    });
  return server;
}

export { createApolloServer };
