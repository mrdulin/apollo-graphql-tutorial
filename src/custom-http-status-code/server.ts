import { graphqlExpress } from './expressApollo';
import express from 'express';
import { makeExecutableSchema } from 'graphql-tools';
import { gql } from 'apollo-server';

const app = express();
const port = 3000;
const graphqlEndpoint = '/graphql';

const typeDefs = gql`
  type Query {
    _: String
  }
`;
const resolvers = {
  Query: {
    _: () => {
      throw new Error('some error');
    },
  },
};
const schema = makeExecutableSchema({ typeDefs, resolvers });

app.use(express.json());
app.use(graphqlEndpoint, graphqlExpress({ schema }));

app.use((error, req, res, next) => {
  if (error === 'some error') {
    return res.status(400).send({ message: error.message ? error.message : error });
  }
  res.status(500);
  res.render('error', { error });
});

app.listen(port, () => {
  console.log(`Http server is listening on http://localhost:${port}${graphqlEndpoint}`);
});
