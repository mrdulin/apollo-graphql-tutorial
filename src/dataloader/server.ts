import { ApolloServer, ServerInfo } from 'apollo-server';
import { typeDefs } from './typeDefs';
import { resolvers } from './resolvers';
import { db } from './db';
import { ApolloServerBase } from 'apollo-server-core';
import { UserLoader, User } from './modules/user';

async function createApolloServer(): Promise<ApolloServerBase> {
  const PORT = process.env.PORT || 3000;
  const server = new ApolloServer({ typeDefs, resolvers, context: { db, UserLoader, User } });

  await server
    .listen(PORT)
    .then(({ url }: ServerInfo) => {
      console.log(`ðŸš€ Server ready at ${url}`);
    })
    .catch((error) => {
      console.error('Create server failed.');
      console.error(error);
    });
  return server;
}

export { createApolloServer };
