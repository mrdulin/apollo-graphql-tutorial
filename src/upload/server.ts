import { ApolloServer } from 'apollo-server';
import { typeDefs } from './typeDefs';
import { resolvers } from './resolvers';
import { ApolloServerBase, FileUploadOptions } from 'apollo-server-core';

async function createApolloServer(): Promise<ApolloServerBase> {
  const fileUploadOptions: FileUploadOptions = {
    // 50KB
    maxFileSize: 1024 * 50,
  };
  const server = new ApolloServer({
    uploads: fileUploadOptions,
    typeDefs,
    resolvers,
  });

  try {
    const { url } = await server.listen();
    console.log(`ðŸš€ Server ready at ${url}`);
    return server;
  } catch (error) {
    throw error;
  }
}

export { createApolloServer };
