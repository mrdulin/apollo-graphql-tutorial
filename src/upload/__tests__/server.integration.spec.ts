import { createTestClient } from 'apollo-server-testing';
import { createApolloServer } from '../server';
import fetch from 'node-fetch';
import FormData from 'form-data';
import fs from 'fs';
import path from 'path';
import { ApolloServerBase } from 'apollo-server-core';

// https://github.com/jaydenseric/graphql-multipart-request-spec
// https://github.com/jaydenseric/graphql-upload/blob/master/src/test.mjs

let server: ApolloServerBase;
let testClient;
beforeAll(async () => {
  server = await createApolloServer();
  testClient = createTestClient(server);
});

afterAll(async () => {
  await server.stop();
});

async function upload(body) {
  return fetch('http://localhost:4000', { method: 'POST', body }).then((response) => response.json());
}

describe('upload', () => {
  it('should upload file correctly with test server', async () => {
    const body = new FormData();
    body.append(
      'operations',
      JSON.stringify({
        query: `
          mutation ($file: Upload!) {
            singleUpload(file: $file) {
              code
              message
            }
          }
        `,
        variables: {
          file: null,
        },
      }),
    );

    body.append('map', JSON.stringify({ 1: ['variables.file'] }));
    body.append('1', 'a', { filename: 'a.txt' });
    const json = await upload(body);
    expect(json).toMatchSnapshot();
  });

  it('should get file size limit error', async () => {
    const body = new FormData();
    body.append(
      'operations',
      JSON.stringify({
        query: `
          mutation ($file: Upload!) {
            singleUpload(file: $file) {
              code
              message
            }
          }
        `,
        variables: {
          file: null,
        },
      }),
    );

    body.append('map', JSON.stringify({ 1: ['variables.file'] }));
    body.append('1', 'b'.repeat(70000), { filename: 'b.txt' });
    const json = await upload(body);
    expect(json).toMatchSnapshot();
  });
});
