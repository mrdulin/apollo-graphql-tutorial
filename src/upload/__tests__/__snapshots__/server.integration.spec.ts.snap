// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`upload should get file size limit error 1`] = `
Object {
  "data": null,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": Object {
          "message": "File truncated as it exceeds the size limit.",
        },
      },
      "locations": Array [
        Object {
          "column": 13,
          "line": 3,
        },
      ],
      "message": "File truncated as it exceeds the size limit.",
      "path": Array [
        "singleUpload",
      ],
    },
  ],
}
`;

exports[`upload should upload file correctly with test server 1`] = `
Object {
  "data": Object {
    "singleUpload": Object {
      "code": 0,
      "message": "",
    },
  },
}
`;
